src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -105,32 +101,44 @@ internal static unsafe bool IsValid(char* name, int start, ref int end, bool all
         //                 / "2" %x30-34 DIGIT     ; 200-249
         //                 / "25" %x30-35          ; 250-255
         //
-        internal static unsafe bool IsValidCanonical(char* name, int start, ref int end, bool allowIPv6, bool notImplicitFile)
+        internal static bool IsValidCanonical<TChar>(ReadOnlySpan<TChar> name, out int bytesConsumed, bool allowIPv6, bool notImplicitFile)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
             int dots = 0;
             int number = 0;
             bool haveNumber = false;
             bool firstCharIsZero = false;
+            int current;
+
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
 
-            while (start < end)
+            bytesConsumed = 0;
+            for (current = 0; current < name.Length; current++)
             {
-                char ch = name[start];
+                TChar ch = name[current];
+
                 if (allowIPv6)
                 {
-                    // for ipv4 inside ipv6 the terminator is either ScopeId, prefix or ipv6 terminator
-                    if (ch == ']' || ch == '/' || ch == '%')
+                    // For an IPv4 address nested inside an IPv6, the terminator is either ScopeId ('%'), prefix ('/') or ipv6 address terminator (']')
+                    if (ch == TChar.CreateChecked('%') || ch == TChar.CreateChecked('/') || ch == TChar.CreateChecked(']'))
+                    {
                         break;
+                    }
                 }
-                else if (ch == '/' || ch == '\\' || (notImplicitFile && (ch == ':' || ch == '?' || ch == '#')))
+                // For a normal IPv4 address, the terminator is the prefix ('/' or its counterpart, '\'). If notImplicitFile is set, the terminator
+                // is one of the characters which signify the start of the rest of the URI - the port number (':'), query string ('?') or fragment ('#')
+                else if (ch == TChar.CreateChecked('/') || ch == TChar.CreateChecked('\\')
+                    || (notImplicitFile && (ch == TChar.CreateChecked(':') || ch == TChar.CreateChecked('?') || ch == TChar.CreateChecked('#'))))
                 {
                     break;
                 }
 
-                if (char.IsAsciiDigit(ch))
+                if (IPAddressParser.TryParseInteger(IPAddressParser.Decimal, ch, out int parsedCharacter))
```

Comment:
Have you benchmarked that extracting this to a helper like `TryParseInteger` is optimal?
E.g. the helper is going through `HexConverter.FromChar` whereas it would have previously been a simple `value - '0'`.

Reply:
I haven't benchmarked it, but I can't see how it could have a measurable performance impact.

The original version would call `char.IsAsciiDigit` followed by `value - '0'`. The new version calls `IPAddressParser.IsValidInteger` and `HexConverter.FromChar`.

`IsValidInteger` will flatten to a range comparison, roughly equivalent to `char.IsAsciiDigit`. `HexConverter.FromChar` is a range check and lookup in the `CharToHex` embedded ReadOnlySpan; the method is marked as inlined. I'd essentially be benchmarking the difference between `value - '0'` and `CharToHexLookup[value]`...